### PR TITLE
Ci/Cd for padl

### PR DESCRIPTION
**Pull request by jasonkhadka**
## Description

Fixes https://github.com/lf1-io/padl/issues/268

Adds pytesting for PRs. 

In next rounds:
- linting
- type checks

### CODE REVIEW

File: .github/workflows/actions.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -r requirements.txt
+        pip install -r requirements-test.txt
+    - name: Test with pytest
+      run: |
+        python -m pytest --cov=padl -n 2
```
Comment (sjrl) [inline .github/workflows/actions.yml:36]: Does the coverage get logged somewhere?

File: requirements.txt [hunk 2]
```diff
@@ -2,3 +2,4 @@ torch
 tqdm
 numpy
 astunparse
+pytest
```
Comment (sjrl) [inline requirements.txt:5]: I don't think `pytest` needs to be in here right?

File: .github/workflows/actions.yml (hunk 1 above)
Comment (sjrl) [reply .github/workflows/actions.yml:36]: I just found it under the actions tab: https://github.com/lf1-io/padl/runs/4262619217?check_suite_focus=true

**Comment by sjrl** [comment]
For future rounds, it might be worth looking into caching dependencies to speed up our tests: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-nodejs-or-python?langId=py#caching-dependencies

File: .github/workflows/actions.yml (hunk 1 above)
Comment (jasonkhadka) [reply .github/workflows/actions.yml:36]: yeah, it is printed on the logs of test step, but not stored in file.

**Comment by sjrl** [comment]
In issue #268 I also brought up the possibility of using an action to rebuild the docs. Would that be possible? I think currently we run the risk of having our docs be out of date with the current source code.

File: requirements.txt (hunk 2 above)
Comment (jasonkhadka) [reply requirements.txt:5]: yeah true, i was testing there and forgot to remove it.

**Comment by jasonkhadka** [comment]
yeah, i think those could be done. I will add them for next updates to actions. 
For now, we dont have a big list of dependencies, so it is farily quick to run the workflow.

> In issue #268 I also brought up the possibility of using an action to rebuild the docs. Would that be possible? I think currently we run the risk of having our docs be out of date with the current source code.

**Comment by sjrl** [comment]
> yeah, i think those could be done. I will add them for next updates to actions. For now, we dont have a big list of dependencies, so it is farily quick to run the workflow.

Do you know if we have "unlimited" computation time for these actions?

**Comment by sjrl** [comment]
@jasonkhadka Also it may be worth adding `pytest --doctest-modules` to our testing to make sure all examples in our docstrings pass. What do you think?

**Comment by jasonkhadka** [comment]
> > yeah, i think those could be done. I will add them for next updates to actions. For now, we dont have a big list of dependencies, so it is farily quick to run the workflow.
> 
> Do you know if we have "unlimited" computation time for these actions?

yes, seems to be unlimited for public repos.

**Comment by sjrl** [comment]
> @jasonkhadka Also it may be worth adding `pytest --doctest-modules` to our testing to make sure all examples in our docstrings pass. What do you think?

Ehh I'm not sure about this at the moment. A lot of them don't pass and I'm not sure if we should spend the effort to get them to work right now.

**Comment by jasonkhadka** [comment]
> > @jasonkhadka Also it may be worth adding `pytest --doctest-modules` to our testing to make sure all examples in our docstrings pass. What do you think?
> 
> Ehh I'm not sure about this at the moment. A lot of them don't pass and I'm not sure if we should spend the effort to get them to work right now.

yeah, dont think this is needed right now. Might be considered while re-writing docs.

**Comment by wuhu** [comment]
> > @jasonkhadka Also it may be worth adding `pytest --doctest-modules` to our testing to make sure all examples in our docstrings pass. What do you think?
> 
> Ehh I'm not sure about this at the moment. A lot of them don't pass and I'm not sure if we should spend the effort to get them to work right now.

Agree to that, many of the examples are not written with doctest in mind (rather they are just for illustration), for example they make reference to things that don't exist like this:

```
>>> mytransform(mysamplething)
blu
```

**Comment by jasonkhadka** [comment]
> > > @jasonkhadka Also it may be worth adding `pytest --doctest-modules` to our testing to make sure all examples in our docstrings pass. What do you think?
> > 
> > 
> > Ehh I'm not sure about this at the moment. A lot of them don't pass and I'm not sure if we should spend the effort to get them to work right now.
> 
> Agree to that, many of the examples are not written with doctest in mind (rather they are just for illustration), for example they make reference to things that don't exist like this:
> 
> ```
> >>> mytransform(mysamplething)
> blu
> ```

just added a small comment in issue #205 (docs related) to think about it, but yeah, would make examples writing bit more complicated.

**Comment by sjrl** [comment]
> > > @jasonkhadka Also it may be worth adding `pytest --doctest-modules` to our testing to make sure all examples in our docstrings pass. What do you think?
> > 
> > 
> > Ehh I'm not sure about this at the moment. A lot of them don't pass and I'm not sure if we should spend the effort to get them to work right now.
> 
> Agree to that, many of the examples are not written with doctest in mind (rather they are just for illustration), for example they make reference to things that don't exist like this:
> 
> ```
> >>> mytransform(mysamplething)
> blu
> ```

@wuhu @jasonkhadka I actually think it might be worth getting these to pass. I'm already going through the doctests now to see if I can get them to pass. I've been making good progress and I may have caught a bug in one of the `dumptools.sourceget` functions.

**Comment by jasonkhadka** [comment]
I think its better to take this discussion on the issue regarding documentation : https://github.com/lf1-io/padl/issues/205
we can list all the problems of examples, and perhaps discuss on writing new examples for those there. 

> > > > @jasonkhadka Also it may be worth adding `pytest --doctest-modules` to our testing to make sure all examples in our docstrings pass. What do you think?
> > > 
> > > 
> > > Ehh I'm not sure about this at the moment. A lot of them don't pass and I'm not sure if we should spend the effort to get them to work right now.
> > 
> > 
> > Agree to that, many of the examples are not written with doctest in mind (rather they are just for illustration), for example they make reference to things that don't exist like this:
> > ```
> > >>> mytransform(mysamplething)
> > blu
> > ```
> 
> @wuhu @jasonkhadka I actually think it might be worth getting these to pass. I'm already going through the doctests now to see if I can get them to pass. I've been making good progress and I may have caught a bug in one of the `dumptools.sourceget` functions.



> > > > @jasonkhadka Also it may be worth adding `pytest --doctest-modules` to our testing to make sure all examples in our docstrings pass. What do you think?
> > > 
> > > 
> > > Ehh I'm not sure about this at the moment. A lot of them don't pass and I'm not sure if we should spend the effort to get them to work right now.
> > 
> > 
> > Agree to that, many of the examples are not written with doctest in mind (rather they are just for illustration), for example they make reference to things that don't exist like this:
> > ```
> > >>> mytransform(mysamplething)
> > blu
> > ```
> 
> @wuhu @jasonkhadka I actually think it might be worth getting these to pass. I'm already going through the doctests now to see if I can get them to pass. I've been making good progress and I may have caught a bug in one of the `dumptools.sourceget` functions.

**Review by sjrl — APPROVED** [review]
Looks good to me!

**Review by wuhu — COMMENTED** [review]
neat!

**Pull request merged**